Source/JavaScriptCore/wasm/WasmTypeDefinition.h

```diff
@@ -92,58 +92,93 @@ enum Mutability : uint8_t {
-    const StructField* storage(StructFieldCount i) const { return const_cast<StructType*>(this)->storage(i); }
+    FieldType& getField(StructFieldCount i) { ASSERT(i < fieldCount()); return *storage(i); }
+    FieldType* storage(StructFieldCount i) { return i + m_payload; }
+    const FieldType* storage(StructFieldCount i) const { return const_cast<StructType*>(this)->storage(i); }
 
 private:
-    StructField* m_payload;
+    FieldType* m_payload;
     StructFieldCount m_fieldCount;
 };
 
+class ArrayType {
+public:
+    ArrayType(FieldType* payload)
+        : m_payload(payload)
+    {
+    }
+
+    FieldType elementType() const { return const_cast<ArrayType*>(this)->getElementType(); }
+
+    WTF::String toString() const;
+    void dump(WTF::PrintStream& out) const;
+
+    FieldType& getElementType() { return *storage(); }
+    FieldType* storage() { return m_payload; }
+    const FieldType* storage() const { return const_cast<ArrayType*>(this)->storage(); }
+
+private:
+    FieldType* m_payload;
+};
+
+enum class TypeDefinitionKind : uint8_t {
+    FunctionSignature,
+    StructType,
+    ArrayType
+};
+
 class TypeDefinition : public ThreadSafeRefCounted<TypeDefinition> {
     WTF_MAKE_FAST_ALLOCATED;
 
     TypeDefinition() = delete;
     TypeDefinition(const TypeDefinition&) = delete;
 
-    TypeDefinition(FunctionArgCount retCount, FunctionArgCount argCount)
+    TypeDefinition(TypeDefinitionKind kind, FunctionArgCount retCount, FunctionArgCount argCount)
         : m_typeHeader { FunctionSignature { static_cast<Type*>(payload()), argCount, retCount } }
     {
+        RELEASE_ASSERT(kind == TypeDefinitionKind::FunctionSignature);
+    }
+
+    TypeDefinition(TypeDefinitionKind kind, StructFieldCount fieldCount)
+        : m_typeHeader { StructType { static_cast<FieldType*>(payload()), fieldCount } }
+    {
+        RELEASE_ASSERT(kind == TypeDefinitionKind::StructType);
     }
 
-    TypeDefinition(StructFieldCount fieldCount)
-        : m_typeHeader { StructType { static_cast<StructField*>(payload()), fieldCount } }
+    TypeDefinition(TypeDefinitionKind kind)
+        : m_typeHeader { ArrayType { static_cast<FieldType*>(payload()) } }
     {
+        RELEASE_ASSERT(kind == TypeDefinitionKind::ArrayType);
     }
 
     // Payload starts past end of this object.
     void* payload() { return this + 1; }
 
     static size_t allocatedFunctionSize(Checked<FunctionArgCount> retCount, Checked<FunctionArgCount> argCount) { return sizeof(TypeDefinition) + (retCount + argCount) * sizeof(Type); }
-    static size_t allocatedStructSize(Checked<StructFieldCount> fieldCount) { return sizeof(TypeDefinition) + fieldCount * sizeof(StructField); }
+    static size_t allocatedStructSize(Checked<StructFieldCount> fieldCount) { return sizeof(TypeDefinition) + fieldCount * sizeof(FieldType); }
+    static size_t allocatedArraySize() { return sizeof(TypeDefinition) + sizeof(FieldType); }
```

Comment:
This is just the length + the pointer to the backing store?

Reply:
For arrays, it's the pointer to the backing store and also the FieldType. Though it's slightly larger from the variant type taking up more than the pointer.

Looking at this again, I'm wondering if it's worth having `ArrayType` use a payload pointer since it doesn't have a variable length of data to store. It possibly saves a tiny amount of space in the `m_typeHeader` in `TypeDefinition` because a `FieldType` is slightly bigger than the size of a `FuncSignature`, so having `ArrayType` store a `FieldType` directly would grow the `std::variant` size slightly.

I will keep it as-is for now though.